File: fpc/bufdither_fpc/ditherproc.pas
unit ditherproc;

{$mode objfpc}{$H+}

interface

uses
  sysutils,
  bufimg;

type
  TPixelFormat = (pf4444, pf565, pf5551);

  TDowngradeProc = function(a, cNum : integer) : integer;

  { TColorReducer }

  TColorReducer = class
    public
      constructor Create(pf : TPixelFormat);
      procedure reduceToClosest(var rgba : TIntRGBA; var destRGBA : TIntRGBA);

    private
      downgr : TDowngradeProc;
  end;


  { TPixelDither }

  TPixelDither = class
    public
      constructor Create;
      destructor Destroy; override;

      procedure ditherImage(img : TBufImgBase; cr : TColorReducer);

    private
      rgbaDiff, rgbaTemp : TIntRGBA;
      img : TBufImgBase;

      procedure calcDiff(var rgba : TIntRGBA; var rgbaReduced : TIntRGBA);
      procedure adjustTemp(coef : integer);inline;
      procedure correctPixel(x, y, coef : integer);
  end;

implementation
var
  down4lookup : array[0..255] of integer;

// most of the time spent in this function (by profiler)
function downgrade(a, targetBitCount : integer) : integer;inline;
var
  maxv : integer;
begin
  maxv := ((1 shl targetBitCount) - 1);
  result := (((a * maxv) div 255) * 255) div maxv;
end;

function downgrade4(a : integer) : integer;inline;
begin
  result := a * 15 div 255 * 17;
end;

function downgrade4444(a, cNum : integer) : integer;
begin
  //result := downgrade(a, 4);
  //result := downgrade4(a);
  result := down4lookup[a];
end;

function downgrade565(a, cNum : integer) : integer;
begin
  if cNum = 1 then
     exit(downgrade(a, 6));

  result := downgrade(a, 5);
end;

function downgrade5551(a, cNum : integer) : integer;
begin
  if cNum = 3 then
     exit(downgrade(a, 1));

  result := downgrade(a, 5);
end;

{ TPixelDither }

constructor TPixelDither.Create;
begin
  inherited;

  img := nil;
end;

destructor TPixelDither.Destroy;
begin
  img := nil;

  inherited Destroy;
end;

procedure TPixelDither.ditherImage(img: TBufImgBase; cr: TColorReducer);
var
  lastCol, lastRow : integer;
  ofs : longint;
  rgba, rgbaReduced : TIntRGBA;
  x, y : integer;
  notLastRow, notLastCol : boolean;
begin
  self.img := img;

  lastCol := img.getWidth-1;
  lastRow := img.getHeight-1;

  for y := 0 to lastRow do
      for x := 0 to lastCol do
          begin
            ofs := img.getOfs(x, y);
            img.getPixelAt(ofs, rgba);
            cr.reduceToClosest(rgba, rgbaReduced);
            img.setPixelAt(ofs, rgbaReduced);

            calcDiff(rgba, rgbaReduced);

            //////////////////////////
            // order, apply error to original pixels
            // (x-1,y+1) = 3/16 , (x,y+1) = 5/16, (x+1,y+1) = 1/16, (x+1, y)=7/16

            notLastRow := (y < lastRow);
            notLastCol := (x < lastCol);
            if notLastRow then
               begin
                    if x > 0 then
                       correctPixel(x-1, y+1, 3);

                    correctPixel(x, y+1, 5);

                    if notLastCol then
                       correctPixel(x+1, y+1, 1);
               end;

            if notLastCol then
               correctPixel(x+1, y, 7);
          end;


  self.img := nil;
end;

function clamp(v : integer) : integer;inline;
begin
  if v < 0 then
     exit(0);

  if v > 255 then
     exit(255);

  result := v;
end;

procedure TPixelDither.calcDiff(var rgba: TIntRGBA; var rgbaReduced: TIntRGBA);
var
  i : integer;
begin
  for i := 0 to 3 do
      rgbaDiff[i] := rgba[i] - rgbaReduced[i];
end;

// secon most time consuming function according to the profile data
procedure TPixelDither.adjustTemp(coef: integer);inline;
var
  i : integer;
begin
  for i := 0 to 3 do
      begin
        rgbaTemp[i] := clamp(rgbaTemp[i] + rgbaDiff[i] * coef div 16);
      end;
end;

procedure TPixelDither.correctPixel(x, y, coef: integer);
var
  ofs : longint;
begin
  //if img.isInBounds(x, y) then
  //   begin
       ofs := img.getOfs(x, y);
       img.getPixelAt(ofs, rgbaTemp);
       adjustTemp(coef);
       img.setPixelAt(ofs, rgbaTemp);
   //  end;

end;


{ TColorReducer }

constructor TColorReducer.Create(pf: TPixelFormat);
begin
  inherited Create;
  downgr := nil;

  case pf of
  pf4444 : downgr := @downgrade4444;
  pf565 : downgr := @downgrade565;
  pf5551 : downgr := @downgrade5551;
  end;
end;

procedure TColorReducer.reduceToClosest(var rgba: TIntRGBA;
  var destRGBA: TIntRGBA);
var
  i : integer;
begin
  for i := 0 to 3 do
      destRGBA[i] := downgr(rgba[i], i);
end;


procedure initlookup4;
var
  i : integer;
begin
  for i := 0 to 255 do
      down4lookup[i] := downgrade4(i);
end;

initialization

initlookup4;

end.

File: nimrod/ditherproc.nim
import bufimg
import strutils

type
    TPixelFormat* = enum
        pf4444, pf565, pf5551
        
    TDowngradeProc = proc(a, cNum : int32): int32 {.noconv.}
    
    TColorReducer* = object of TObject
        downgr: TDowngradeProc
        
    TPixelDither* = object of TObject
        rgbaDiff, rgbaTemp: TRGBA
        img: ref TBufImgBase
        
        
# ------ TColorReducer
var
  down4lookup: array[0..255, int32]


proc downgrade4444(a, cNum : int32) : int32 {.noconv.} =
    result = down4lookup[a]


proc newColorReducer*(pf : TPixelFormat) : ref TColorReducer =
    new(result)
    if pf == pf4444:
        result.downgr = downgrade4444
    else:
        quit "only pf4444 is supported!"
    
     
proc reduceToClosest*(red: ref TColorReducer, rgba: var TRGBA, reduced: var TRGBA)  =
    for i in countup(0, 3'i32):
        reduced[i] = red.downgr(rgba[i], i)
    

# ------ TPixelDither

proc newPixelDither*(): ref TPixelDither =
    new(result)
    
    
proc clamp(v : int32) : int32 {.inline.} =
    if v < 0:
        return(0)

    if v > 255:
        return(255)

    result = v


proc calcDiff(pd: ref TPixelDither, rgba: var TRGBA, rgbaReduced: var TRGBA) =
  for i in countup(0,3):
      pd.rgbaDiff[i] = rgba[i] - rgbaReduced[i]


proc adjustTemp(pd: ref TPixelDither, coef: int32) =
  for i in countup(0,3):
    pd.rgbaTemp[i] = clamp(pd.rgbaTemp[i] + pd.rgbaDiff[i] * coef div 16)


proc correctPixel(pd: ref TPixelDither, x, y, coef: int32) =
   #if pd.img.inBounds(x,y):
   let ofs = pd.img.ofs(x, y)
   pd.img.getpixel(ofs, pd.rgbaTemp)
   pd.adjustTemp(coef)
   pd.img.setpixel(ofs, pd.rgbaTemp)

  
when false:  
    proc dumpRGBA(rgba: TRGBA) =
        stdout.write("(" & inttostr(rgba[0]) & ", " & inttostr(rgba[1]) & ", " & inttostr(rgba[2]) & ", " & inttostr(rgba[3]) & ")")
    
    
proc ditherImage*(pd: ref TPixelDither, img: ref TBufImgBase, cr: ref TColorReducer) =
    pd.img = img
    let lastCol = img.width-1
    let lastRow = img.height-1
    
    for y in countup(0, lastRow):
        for x in countup(0, lastCol):
            let ofs = img.ofs(x,y)
            var rgba: TRGBA
            img.getPixel(ofs, rgba)
            var rgbaReduced: TRGBA
            cr.reduceToClosest(rgba, rgbaReduced)
            
            #dumpRGBA(rgba)
            #stdout.write(" -> ")
            #dumpRGBA(rgbaReduced)
            #stdout.writeln("") 
            
            img.setpixel(ofs, rgbaReduced)
            
            pd.calcDiff(rgba, rgbaReduced)
    
            # order, apply error to original pixels
            # (x-1,y+1) = 3/16 , (x,y+1) = 5/16, (x+1,y+1) = 1/16, (x+1, y)=7/16

            when true:

                let notLastRow = (y < lastRow)
                let notLastCol = (x < lastCol)
                
                
                if notLastRow:
                    if x > 0:
                        pd.correctPixel(x-1, y+1, 3)

                    pd.correctPixel(x, y+1, 5)

                    if notLastCol:
                       pd.correctPixel(x+1, y+1, 1)

                if notLastCol:
                   pd.correctPixel(x+1, y, 7)
               
            else:
                pd.correctPixel(x-1, y+1, 3)
                pd.correctPixel(x, y+1, 5)
                pd.correctPixel(x+1, y+1, 1)
                pd.correctPixel(x+1, y, 7)

    pd.img = nil
    
    

# -----

#function downgrade(a, targetBitCount : integer) : integer;inline;
#var
#  maxv : integer;
#begin
#  maxv := ((1 shl targetBitCount) - 1);
#  result := (((a * maxv) div 255) * 255) div maxv;
#end;

proc downgrade4(a : int32) : int32 =
  result = a * 15 div 255 * 17

proc initlookup4 = 
    for i in countup(0, 255'i32):
        down4lookup[i] = downgrade4(i)
        #stdout.writeln($i & " -> " & $down4lookup[i])
    #stdout.writeln "initialized lookup"
          
# -------
initlookup4()

